rust_src/src/windows.rs

```diff
@@ -1536,4 +1592,131 @@ pub fn window_body_width_lisp(window: LispWindowLiveOrSelected, pixelwise: bool)
+/// first line of WINDOW.
+///
+/// LAST, if non-nil, specifies the last line whose dimensions shall be
+/// returned.  If LAST is nil and BODY is non-nil, the last line is the last
+/// line of the body (text area) of WINDOW.  Otherwise, last is the last
+/// line of WINDOW.
+///
+/// INVERSE, if nil, means that the y-pixel value returned for a specific
+/// line specifies the distance in pixels from the left edge (body edge if
+/// BODY is non-nil) of WINDOW to the right edge of the last glyph of that
+/// line.  INVERSE non-nil means that the y-pixel value returned for a
+/// specific line specifies the distance in pixels from the right edge of
+/// the last glyph of that line to the right edge (body edge if BODY is
+/// non-nil) of WINDOW.
+///
+/// LEFT non-nil means to return the x- and y-coordinates of the lower left
+/// corner of the leftmost character on each line.  This is the value that
+/// should be used for buffers that mostly display text from right to left.
+///
+/// If LEFT is non-nil and INVERSE is nil, this means that the y-pixel value
+/// returned for a specific line specifies the distance in pixels from the
+/// left edge of the last (leftmost) glyph of that line to the right edge
+/// (body edge if BODY is non-nil) of WINDOW.  If LEFT and INVERSE are both
+/// non-nil, the y-pixel value returned for a specific line specifies the
+/// distance in pixels from the left edge (body edge if BODY is non-nil) of
+/// WINDOW to the left edge of the last (leftmost) glyph of that line.
+///
+/// Normally, the value of this function is not available while Emacs is
+/// busy, for example, when processing a command.  It should be retrievable
+/// though when run from an idle timer with a delay of zero seconds.
+#[lisp_fn]
+pub fn window_lines_pixel_dimensions(
+    window: LispWindowLiveOrSelected,
+    first: Option<LispNumber>,
+    last: Option<LispNumber>,
+    body: bool,
+    inverse: bool,
+    left: bool,
+) -> LispObject {
+    let window: LispWindowRef = window.into();
+    let buffer: LispBufferRef = window.contents.into();
+    let max_y = if body {
+        window.text_bottom_y()
+    } else {
+        window.pixel_height
+    };
+    let window_width = if body {
+        window.body_width(true)
+    } else {
+        window.pixel_width
+    };
+    let header_line_height = window.header_line_height();
+    let subtract = if body { header_line_height } else { 0 };
+    let matrix = window.get_matrix();
+
+    if unsafe { noninteractive } || window.is_pseudo() {
+        return Qnil;
+    }
+
+    // Fail if current matrix is not up to date.
+    unsafe {
+        if !window.window_end_valid()
+            || windows_or_buffers_changed > 0
+            || buffer.clip_changed()
+            || buffer.prevent_redisplay_optimizations_p()
+            || window.is_outdated()
+        {
+            return Qnil;
+        }
+    }
+
+    let mut row = if let Some(first) = first {
```

Comment:
Why use `if let Some() ... else ...` instead of a `match`? I find that when there is an else, the `match` form is a better fit.

Reply:
I was on the fence on which to use, but after rewriting it I can see `match` is the nicer option